collect_app/src/main/java/org/odk/collect/android/adapters/InstanceUploaderAdapter.java

```diff
@@ -0,0 +1,264 @@
+            }
+
+            SmsEvent currentStatus = new SmsEvent();
+            currentStatus.setStatus(messageStatus);
+            currentStatus.setLastUpdated(model.getLastUpdated());
+            currentStatus.setProgress(model.getCompletion());
+
+            setDisplaySubTextView(currentStatus, viewHolder);
+
+            setupCloseButton(viewHolder, messageStatus);
+            viewHolder.closeButton.setOnClickListener(v -> smsService.cancelFormSubmission(String.valueOf(instanceId)));
+        }
+
+        compositeDisposable.add(eventBus.register(SmsEvent.class)
+                .filter(event -> event.getInstanceId().equals(String.valueOf(instanceId)))
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(event -> {
+                    viewHolder.progressBar.setProgressPercent((int) event.getProgress().getPercentage(), true);
+                    setSmsSubmissionStateIcons(event.getStatus(), viewHolder);
+                    setDisplaySubTextView(event, viewHolder);
+                    setupCloseButton(viewHolder, event.getStatus());
+                }));
+    }
+
+    private void setupCloseButton(ViewHolder viewHolder, MessageStatus status) {
+
+        if (status == null) {
+            return;
+        }
+
+        if (status.equals(MessageStatus.Sending)) {
+            viewHolder.closeButton.setVisibility(View.VISIBLE);
+            viewHolder.checkbox.setVisibility(View.GONE);
+        } else {
+            viewHolder.closeButton.setVisibility(View.GONE);
+            viewHolder.checkbox.setVisibility(View.VISIBLE);
+        }
+    }
+
+    private void setSmsSubmissionStateIcons(MessageStatus messageStatus, ViewHolder viewHolder) {
+
+        if (messageStatus == null) {
+            return;
+        }
+
+        switch (messageStatus) {
+            case Delivered:
+            case Sent:
+                viewHolder.statusIcon.setImageResource(R.drawable.check);
+                break;
+
+            case Queued:
+            case Sending:
+                viewHolder.statusIcon.setImageResource(R.drawable.message_text_outline);
+                break;
+
+            case Ready:
+                viewHolder.statusIcon.setImageResource(R.drawable.pencil);
+                break;
+
+            default:
+                viewHolder.statusIcon.setImageResource(R.drawable.message_alert);
+                break;
+        }
+    }
+
+    private void setDisplaySubTextView(SmsEvent progress, ViewHolder viewHolder) {
+        String text = getDisplaySubtext(progress);
+        if (text != null) {
+            viewHolder.displaySubtext.setText(text);
+        }
+    }
+
+    private String getDisplaySubtext(SmsEvent event) {
+        Date date = event.getLastUpdated();
+        SmsProgress progress = event.getProgress();
+
+        if (event.getStatus() == null) {
+            return null;
+        }
+
+        switch (event.getStatus()) {
```

Comment:
This switch statement should catch IllegalArgumentException we had a similar problem in the past #1774 you know strings are added by users using Transifex and some of them might be wrong and cause exceptions so please add it just in case.